cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java

```diff
@@ -100,8 +106,27 @@ public void set(BasicMapReduceContext context, CConfiguration conf, Transaction
     setPlugins(context.getApplicationSpecification().getPlugins());
     setProgramJarURI(programJarURI);
     setConf(conf);
-    setTx(tx);
     setLocalizedResources(localizedUserResources);
+    setOutputs(context.getOutputs());
+  }
+
+  private void setOutputs(List<ProvidedOutput> providedOutputs) {
+    // we only need to serialize the original Output objects, not the entire ProvidedOutput
+    List<Output.DatasetOutput> datasetOutputs = new ArrayList<>();
+    for (ProvidedOutput providedOutput : providedOutputs) {
+      Output output = providedOutput.getOutput();
+      if (output instanceof Output.DatasetOutput) {
```

Comment:
Technically, someone can add an object of the `OutputFormatProvider` interface and also implement the `DatasetOutputCommitter` interface. We used to have the user-provided object available and so we could call the `DatasetOutputCommitter` methods on it in the MapReduceRuntimeService. However, since its an object of the `OutputFormatProvider` interface, we can't know how to serialize it to the MainOutputCommitter (which runs in the MapReduce AM).
Since this is a less common case, the approach I took is to disallow users from using the `addOutput(Output.of(String outputName, OutputFormatProvider outputFormatProvider))` API if their specified OutputFormatProvider is also a DatasetOutputCommitter.